chenyulin0719/yunikorn-k8shim
# [YUNIKORN-2631] Support canonical labels for queue/applicationId in Admission Controller

### What is this PR for?

When Admission Controller try to patch queue/applicationID to pod, also add canonical representations.

- applicationId (legacy, should be removed in 1.7.0)
- queue  (legacy, should be removed in 1.7.0)
- yunikorn.apache.org/app-id **(New, Canonical Label)**
- yunikorn.apache.org/queue  **(New, Canonical Label)**

In 1.6.0, if only canonical labels were found in pod, add the value back to legacy labels for backward compatibility
- applicationId 
- queue 

### What type of PR is it?


### Todos

- Support canonical representation in scheduler
- Update documents

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-2631

### How should this be tested?
1. make test
2. create below pod and check the pod labels
- pod-with-legacy-labels: AM won't add new labels
- pod-with-no-labels: AM add both legacy/canonical labels
- pod-with-canonical-labels: Besides the existing canonical labels, AM also add the value to legacy labels.
```
apiVersion: v1
kind: Pod
metadata:
  labels:
    app: sleep
    applicationId: "application-sleep-0001"
    queue: "root.sandbox"
  name: pod-with-legacy-labels
spec:
  schedulerName: yunikorn
  restartPolicy: Never
  containers:
    - name: sleep-6000s
      image: "alpine:latest"
      command: ["sleep", "6000"]
      resources:
        requests:
          cpu: "100m"
          memory: "500M"
---
apiVersion: v1
kind: Pod
metadata:
  labels:
    app: sleep
  name: pod-with-no-labels
spec:
  schedulerName: yunikorn
  restartPolicy: Never
  containers:
    - name: sleep-6000s
      image: "alpine:latest"
      command: ["sleep", "6000"]
      resources:
        requests:
          cpu: "100m"
          memory: "500M"
---
apiVersion: v1
kind: Pod
metadata:
  labels:
    app: sleep
    yunikorn.apache.org/app-id: "application-sleep-0003"
    yunikorn.apache.org/queue: "root.sandbox"
  name: pod-with-canonical-labels
spec:
  schedulerName: yunikorn
  restartPolicy: Never
  containers:
    - name: sleep-6000s
      image: "alpine:latest"
      command: ["sleep", "6000"]
      resources:
        requests:
          cpu: "100m"
          memory: "500M"


```

### Screenshots (if appropriate)

- pod-with-legacy-labels
<img width="689" alt="image" src="https://github.com/chenyulin0719/yunikorn-k8shim/assets/26764036/559afa8e-a48f-4f36-abbd-ee7281b76f05">

- pod-with-no-labels
<img width="665" alt="image" src="https://github.com/chenyulin0719/yunikorn-k8shim/assets/26764036/63d485d4-6213-4bd6-87e8-17f1dcee7809">


- pod-with-canonical-labels:
<img width="688" alt="image" src="https://github.com/chenyulin0719/yunikorn-k8shim/assets/26764036/7671ce4c-8986-40c6-9b21-2efa70cec2c7">




### Questions:
NA
